### PR TITLE
Release 23.6.0

### LINKED ISSUE #514: Publish using the new PyPI and do a release to test
The details are here

https://docs.pypi.org/trusted-publishers/using-a-publisher/

To actually test this, we will need to trigger a full release 

And with the full release, we can drop support for python 3.7... even if it's not EOL yet.

Byt this will be the last release to support 3.7

### PR DESCRIPTION
**Pull request by adiroiban**
# Description

This is the release PR for 23.6.0

Fixes #514 and while doing this release we will test the new publishing method.

This is also the last release for python 3.7.
This is done to simplify our code.

I am just following the info from RELEASE.rst

### CODE REVIEW

**Comment by adiroiban** [comment]
needs-review

**Comment by hynek** [comment]
Just double-checking: is OIDC already configured for towncrier in PyPI? I don't think I've received and email about that. 🤔

**Comment by adiroiban** [comment]
I am reading the docs and the initial announcement here https://blog.pypi.org/posts/2023-04-20-introducing-trusted-publishers/ and my understanding is that OIDC is already enabled for all repos that are hosted on GitHub.

For now, only GitHub is supported :( ... but we are on GitHub, so this doesn't affects us.

**Comment by hynek** [comment]
It is enabled, but you still have to set some settings.

The project view has look something like this:

CleanShot 2023-06-05 at 12 18 51@2x

If it doesn't, you just have to fill out this form:

CleanShot 2023-06-05 at 12 19 24@2x

**Comment by adiroiban** [comment]
I see... I don't have admin rights to `towncrier` in PyPi... the release just worked with the token from GitHub Secrets.

I will see if I/we can get admin rights for `towncrier` in PyPI

**Comment by hynek** [comment]
@glyph? 😇

File: .github/workflows/ci.yml
Comment (hynek) [inline .github/workflows/ci.yml:198]: It would be good to additionally set an environment (I use `release-pypi` myself) which you then configure on PyPI to be the only one that is allowed to publish:

```suggestion
    environment: release-pypi
    permissions:
```

My understanding is that it gets created automatically when used, but you can also manually add it to https://github.com/twisted/towncrier/settings/environments

Here's a quote from the publisher docs on PyPI:

> The name of the [GitHub Actions environment](https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment) that the above workflow uses for publishing. This should be configured under the repository's settings. While not required, a dedicated publishing environment is strongly encouraged, especially if your repository has maintainers with commit access who shouldn't have PyPI publishing access.

Comment (adiroiban) [reply .github/workflows/ci.yml:198]: Done. I am not sure this is needed.

I hope we can keep the towncrier access security setting simple ... and fully trust anyone with write access to the main repo.

Fork should be ok as the forks should not get access to the GitHub Secrets and I hope PyPI will deny access to the OIDC token from forks :)

Comment (adiroiban) [reply .github/workflows/ci.yml:198]: @hynek it looks like using the GHA environment has the unwanted side effect, that any CI run is now blocked and  needs manual approval.

I tried to add a high-level conditional, but I see that the deploy request is still triggered.

Comment (hynek) [reply .github/workflows/ci.yml:198]: Eh ok leave it out for now and we’ll look another time. Sorry for the noise.

Comment (adiroiban) [reply .github/workflows/ci.yml:198]: I  think that the environment thing is worth exploring ... 

I guess that the deployment workflow is designed to be triggered as a side-effect of another workflow (with dispatch) ... and it's not designed to be executed as mixed job

Comment (hynek) [reply .github/workflows/ci.yml:198]: This is how I do it: https://github.com/hynek/structlog/blob/main/.github/workflows/pypi-package.yml

If you want, you can copy it verbatim and just change the trigger conditions.

**Comment by glyph** [comment]
the twisted core team is now an owner of the towncrier project.

**Comment by adiroiban** [comment]
Thanks @glyph . I have access twisted/towncrier and I have created the GitHub publisher.

But I still don't have access twisted/twisted

**Comment by glyph** [comment]
OK I guess the 'maintainer' role isn't sufficient for most tasks. I promoted 'core' to 'owner' on twisted/twisted

File: .github/workflows/ci.yml
Comment (adiroiban) [reply .github/workflows/ci.yml:198]: I think it make sense to have the release code in a separate workflow.

At the same time, I hope we can keep `towncrier` automation simple and have everything in a single file to make it easier to discover and see what is going on in GitHub Actions.

------------

I don't know what's best here :|

**Comment by adiroiban** [comment]
I think this is ready for review.

The release notes are now visible in HTML format

https://towncrier--515.org.readthedocs.build/en/515/release-notes.html# 

please review

File: NEWS.rst
Comment (hynek) [inline NEWS.rst]: this is not great (and entirely my fault), but I  don't think it's worth to stop the release over

is it possible for you to move it down to the misc list?

File: .github/workflows/ci.yml
Comment (hynek) [reply .github/workflows/ci.yml:198]: I personally think (as demonstrated :)) that CI and publishing are sufficiently distinct issues, that it absolutely makes sense to keep them apart. But that's not something we need / should solve _here_.

**Review by hynek — APPROVED** [review]
There is the small issue of an empty doc changelog entry that's my fault too, but nbd. if you want, fix that, otherwise go for it.

**Comment by adiroiban** [comment]
Yes. We definitely need to move the release/publish in a separate workflow.

The current workflow is confusing.

so ... release candidate is on pypi - https://pypi.org/project/towncrier/23.6.0rc1/

Published here - https://github.com/twisted/towncrier/actions/runs/5188156080/jobs/9351471004

----------

I will prepare the final release

File: .github/workflows/ci.yml
Comment (chrysle) [reply .github/workflows/ci.yml:198]: I would like to experiment with that and make another PR.

**Comment by hynek** [comment]
> so ... release candidate is on pypi - https://pypi.org/project/towncrier/23.6.0rc1/
> 
> Published here - https://github.com/twisted/towncrier/actions/runs/5188156080/jobs/9351471004

Happy to see it worked out!
 
> I will prepare the final release

Thanks for doing this!

**Comment by adiroiban** [comment]
I have created a separate ticket to discuss moving the release in a separate file #519

final release done here https://github.com/twisted/towncrier/actions/runs/5200182732

The package is published https://pypi.org/project/towncrier/23.6.0/

I will now set the dev version and merge this  PR.

Thanks for your help

**Comment by adiroiban** [comment]
I have  removed the `PYPI_TOKEN` secret from here  https://github.com/twisted/towncrier/settings/secrets/actions  ... this is the GHA side.

But I don't know where this token is defined inside PyPI. 
I see that on GHA it was created on Updated on Dec 9, 2020
I can also see the log on PyPI with the token creation by @glyph  - https://pypi.org/manage/project/towncrier/history/?page=2

Maybe the token is hosted in Glyphs PyPI profile.
Glyph, if you can find the Towncrier PyPI token , you can delete it :)

File: NEWS.rst
Comment (chrysle) [inline NEWS.rst:22]: [GitHub release tags](https://github.com/twisted/towncrier/releases/tag/23.6.0) don't seem to be able to render RST links such as this.

Is that a problem?

Comment (hynek) [reply NEWS.rst:22]: I've pandoc'ed it now.

**Pull request merged**
